nomad/csi_volume_endpoint.go

```diff
@@ -0,0 +1,239 @@
+package nomad
+
+import (
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	log "github.com/hashicorp/go-hclog"
+	memdb "github.com/hashicorp/go-memdb"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// CSIVolume wraps the structs.CSIVolume with request data and server context
+type CSIVolume struct {
+	srv    *Server
+	logger log.Logger
+	// ctx    *RPCContext
+}
+
+// QueryACLObj looks up the ACL token in the request and returns the acl.ACL object
+// - fallback to node secret ids
+func (srv *Server) QueryACLObj(args *structs.QueryOptions) (*acl.ACL, error) {
+	// Lookup the token
+	aclObj, err := srv.ResolveToken(args.AuthToken)
+	if err != nil {
+		// If ResolveToken had an unexpected error return that
+		return nil, err
+	}
+
+	if aclObj == nil {
+		ws := memdb.NewWatchSet()
+		node, stateErr := srv.fsm.State().NodeBySecretID(ws, args.AuthToken)
+		if stateErr != nil {
+			// Return the original ResolveToken error with this err
+			var merr multierror.Error
+			merr.Errors = append(merr.Errors, err, stateErr)
+			return nil, merr.ErrorOrNil()
+		}
+
+		if node == nil {
+			return nil, structs.ErrTokenNotFound
+		}
+	}
+
+	return aclObj, nil
+}
+
+// WriteACLObj calls QueryACLObj for a WriteRequest
+func (srv *Server) WriteACLObj(args *structs.WriteRequest) (*acl.ACL, error) {
+	opts := &structs.QueryOptions{
+		Region:    args.RequestRegion(),
+		Namespace: args.RequestNamespace(),
+		AuthToken: args.AuthToken,
+	}
+	return srv.QueryACLObj(opts)
+}
+
+// replyCSIVolumeIndex sets the reply with the last index that modified the table csi_volumes
+func (srv *Server) replyCSIVolumeIndex(state *state.StateStore, reply *structs.QueryMeta) error {
+	// Use the last index that affected the table
+	index, err := state.Index("csi_volumes")
+	if err != nil {
+		return err
+	}
+	reply.Index = index
+
+	// Set the query response
+	srv.setQueryMeta(reply)
+	return nil
+}
+
+// List replies with CSIVolumes, filtered by ACL access
+func (v *CSIVolume) List(args *structs.CSIVolumeListRequest, reply *structs.CSIVolumeListResponse) error {
```

Comment:
General design question: Do we want to pre-emptively paginate this? I could see the volumes list getting large in some clusters, even though most will probably have a smaller set.

Reply:
That's a good call, yes I'd like to paginate by default.

Reply:
Pagination here is a little complex, so I'm going to address it in a followup